#                                                 Assignment - 4
###       Gauthami



In [1]:
import math
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import StratifiedKFold, cross_val_score, cross_val_predict 
from sklearn.model_selection import train_test_split
from sklearn import preprocessing, metrics
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import KFold

For Data analysis, Data Preprocessing and Data Modelling, we have imported all the libraries 

##  <font color='red'>*Data Preprocessing*</font>


### <font color='Blue'>Step1</font>



In [25]:
df =  pd.read_csv(r'C:\Users\skura\Desktop\Classes\Principles of Data Management and Mining\Assign4\breastcancer.csv',delimiter=',',header=None)
df.columns = ['Sample code', 'Clump Thickness', 'Uniformity of Cell Size', 'Uniformity of Cell Shape',
                'Marginal Adhesion', 'Single Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin',
                'Normal Nucleoli', 'Mitoses','Class']

print('Number of instances = %d' % (df.shape[0]))
print('Number of attributes = %d' % (df.shape[1]))
df.head(5)


Number of instances = 699
Number of attributes = 11


Sample code  Clump Thickness  Uniformity of Cell Size  \
0      1000025                5                        1   
1      1002945                5                        4   
2      1015425                3                        1   
3      1016277                6                        8   
4      1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

  Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0           1                3                1        1      2  
1          10                3                2        1      2  
2           2                3                1        1      2  
3           4                3                7        1      2  
4           1                3                1        1      2

We have loaded up data set into a Pandas DataFrame and assigned column names.Below 


No of instances = 699

No of Attributes = 11


### <font color='purple'>Step2</font>

In [26]:
df = df.drop(['Sample code'],axis=1)
df = df.replace('?',np.NaN)

print('Number of instances = %d' % (df.shape[0]))
print('Number of attributes = %d' % (df.shape[1]))



print('Number of missing values:')
for col in df.columns:
    print('\t%s: %d' % (col,df[col].isna().sum()))


# Missing Value  in attribute
print('Number of rows in original data = %d' % (df.shape[0]))
data = df.dropna()
print('Number of rows after discarding missing values = %d' % (data.shape[0]))


Number of instances = 699
Number of attributes = 10
Number of missing values:
	Clump Thickness: 0
	Uniformity of Cell Size: 0
	Uniformity of Cell Shape: 0
	Marginal Adhesion: 0
	Single Epithelial Cell Size: 0
	Bare Nuclei: 16
	Bland Chromatin: 0
	Normal Nucleoli: 0
	Mitoses: 0
	Class: 0
Number of rows in original data = 699
Number of rows after discarding missing values = 683


After removing the Sample code solumns, which has no importance as feature to predict the cancer, we also replaced missing values with questions and removed them. There are **16 missing values** in the dataset. After Cleaning data we got following resultshere are 16 instances in Groups 1 to 6 that contain a single missing 

**Number of rows in original data** = 699

**Bare Nuclei** =16

**Number of rows after discarding missing values** = 683



## <font color='red'>Normalization</font>

In [39]:


from sklearn import preprocessing
# Get column names first
names = data.columns
# Create the Scaler object
scaler = preprocessing.StandardScaler()
encoder = preprocessing.LabelEncoder()
df['Class'] = encoder.fit_transform(df['Class'])
X = np.array(data.drop(['Class'], 1))
y = np.array(data['Class'])

# Fit your data on the scaler object
X = scaler.fit_transform(X)


Using Preprocessing library , we have created a new dataframe that contains the normalized numbers. In order to fit the data into models.


# <font color='red'>Random Forest Classifier</font>

### Use ensemble.RandomForestClassifier with n_estimators=10 and use K-Fold cross validation to get a measure of the accuracy (K=10). Does it perform better than decision tree?

In [63]:
clf = RandomForestClassifier(n_estimators=10, random_state = 36) #Initialize with whatever parameters you want to

# 10-Fold Cross validation

kf =KFold(n_splits = 10, random_state=30, shuffle=False)
RF_accuracy = np.mean(cross_val_score(clf, X, y, cv=kf))



print( "Accuracy for Random Forest Classifier is :{:.2%}".format(RF_accuracy))

Accuracy for Random Forest Classifier is :96.64%


The RandomForestClassifier() from sklearn is a very simple model to build by adding a few parameters to reduce over-fitting. A random forest is a meta estimator that fits a number of decision tree classifiers on various sub-samples of the dataset. Random Forest of **10 decision Trees using K fold cross validation**  to get accuracy of the model which helps determining  whether  breast cancer  is malignant or benign in the class variable.

Accuracy for Random Forest Classifier is :**96.64%**



# <font color='red'>Decision Tree Classifier using K Fold Cross Validation</font>

###  Using  K-Fold cross validation to get a measure of your model’s accuracy (K=10)

In [64]:
decision = DecisionTreeClassifier(criterion = 'entropy', random_state = 0, max_depth=3)
kf =KFold(n_splits = 10, random_state=None, shuffle=False)
tree = np.mean(cross_val_score(decision, X, y, cv=kf))



print( "Accuracy of Decision Tree using K fold Cross Validation :{:.2%}".format(tree))
 

Accuracy of Decision Tree using K fold Cross Validation :94.74%


Accuracy of Decision Tree using K fold Cross Validation : **94.74%**

# <font color='red'>Decision Tree Classifier train and Test Split </font>

In [62]:
X_train, X_test, y_train, y_test = train_test_split(data.loc[:, data.columns != 'Class'],
                                                    data['Class'], stratify= data['Class'], random_state=66, test_size = 0.25, train_size = 0.75)

tree = DecisionTreeClassifier(max_depth=3, criterion = 'entropy')
tree.fit(X_train, y_train)
 

      
print("Decision tree training set accuracy:{:.2%}".format(tree.score(X_train, y_train)))
print("Decision tree testing set accuracy:{:.2%}".format(tree.score(X_test, y_test)))
                                                                    


Decision tree training set accuracy:96.68%
Decision tree testing set accuracy:94.15%


Before we can create the training and test data, we must remove the ‘id’ column. This column provides no value on the prediction on breast cancer and can be removed with the del function from pandas. Using the **Train_Test_Split** function, we have divided data into **75** of training data  and **25%** of test data.

In this DecisionTreeClassifier(), we will add the max_depth parameter which is optional. This parameter sets an integer to the maximum depth of the decision tree. If this left as None, the nodes will be expanded until all leaves are pure or until the leaves contain less than the than min_samples_split samples.


Decision tree training set accuracy  :**96.68%**

Decision tree testing set accuracy   :**94.15%**

#                     <font color='red'>KNN model</font>
 
### Use neighbors.KNeighborsClassifier and use K-Fold cross validation to get a measure of the accuracy (K=10).

In [54]:
accuracy =[]

k = range(1, 50)

for n_neighbors in k:
    clf = KNeighborsClassifier(n_neighbors= n_neighbors)
    kf =KFold(n_splits = 10, random_state=66, shuffle=False)
    knn = np.mean(cross_val_score(clf, X, y, cv=kf))
    accuracy.append(knn)
    
    
print(accuracy)

#print("Max Accuracy  "+str(max(accuracy))+"At Neighbors = "+str(accuracy.index(max(accuracy))+1))



print( "Accuracy for KNeighborsClassifie ne :{:.2%}".format(max(accuracy)) +" " +"At Neighbors = "+str(accuracy.index(max(accuracy))+1))

[0.9488491048593349, 0.9459505541346974, 0.9590792838874682, 0.9605711849957373, 0.9634271099744247, 0.9634697357203752, 0.9663895993179882, 0.9678601875532824, 0.9663469735720376, 0.9649403239556694, 0.9692881500426258, 0.9663895993179882, 0.9663682864450129, 0.9663895993179882, 0.964919011082694, 0.9634910485933504, 0.9634697357203752, 0.9620204603580564, 0.9634697357203752, 0.9664109121909634, 0.9663895993179882, 0.9649616368286444, 0.9649616368286444, 0.9635123614663257, 0.9649616368286444, 0.9635123614663257, 0.9634910485933503, 0.9649616368286444, 0.9634910485933503, 0.9635123614663257, 0.9620417732310316, 0.9620417732310316, 0.9620417732310316, 0.959143222506394, 0.9605924978687128, 0.9605924978687128, 0.9605924978687128, 0.9605924978687128, 0.9605924978687128, 0.9605924978687128, 0.9605924978687128, 0.959143222506394, 0.959143222506394, 0.959143222506394, 0.9605924978687128, 0.959143222506394, 0.959143222506394, 0.959143222506394, 0.959143222506394]
Accuracy for KNeighborsClass

we have  analyzed the Wisconsin breast cancer dataset for prediction using k-nearest neighbors  algorithm. the algorithm draws a perimeter around K and studies the other data points within that perimeter.

From the above results we can view that the nearest **neighbor 11** has the highest  accuracy of **96.93** 

# <font color='red'>Naive Bayes using K Fold Cross Validation</font>

In [37]:
from sklearn.naive_bayes import MultinomialNB

x = np.array(data.drop(['Class'], 1))
scale = preprocessing.MinMaxScaler()
x = scale.fit_transform(x)
kf =KFold(n_splits = 10, random_state=None, shuffle=False)
classifier = MultinomialNB()
model = classifier.fit(x,y)
Naive_Accuracy = np.mean(cross_val_score(model, x, y, cv=kf,scoring='roc_auc'))


print( "Accuracy for Naive Bayes :{:.2%}".format(Naive_Accuracy))
print("Aucroc: %0.3f" % metrics.roc_auc_score(y, cross_val_predict(classifier, x, y, cv=kf)))

Accuracy for Naive Bayes :97.09%
Aucroc: 0.873


Output for a naive Bayes classifier algorithm consists of a data set placed into a number of different categories. Category composition and the speed of the function can also be determined. There is no further quantitative or qualitative analysis performed. That function can then be repeated again and again in order for machine learning to occur.


Accuracy of Naive Bayes is : **97%**

# <font color='red'>Conclusion</font>

Below are summarised accuracies of Different models

Decision tree training set accuracy  :**96.68%**

Decision tree testing set accuracy   :**94.15%**

Accuracy of Decision Tree using K fold Cross Validation :**94.74%**

Accuracy for Random Forest Classifier is :**96.64%**

the K Nearest Neighbor 11 has the highest accuracy  **96.93**

Accuracy of Naive Bayes is : **97%**


**KNN and Naive Bayes**  classfier has the best accuracy compared to other models.We could see that there is no much difference between **Decision Tree**  accuracy and K fold cross validation accuracy. 